src/core/tranche/distributor/distributor.sol

```diff
@@ -0,0 +1,114 @@
+// Copyright (C) 2019 Centrifuge
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+pragma solidity >=0.4.24;
+
+import "ds-note/note.sol";
+
+contract OperatorLike {
+    function borrow(address, uint) public;
+    function debt() public returns (uint);
+    function repay(address, uint) public;
+    function balance() public returns (uint);
+}
+
+contract ManagerLike {
+    function senior() public returns(address);
+    function junior() public returns(address);
+
+    function pile() public returns(address);
+    function poolClosing() public returns(bool);
+
+    function ActionTake() public returns (uint);
+    function ActionGive() public returns (uint);
+    function requestAction() public returns (uint, uint);
+
+}
+
+contract Distributor is DSNote {
+    // --- Auth ---
+    mapping (address => uint) public wards;
+    function rely(address usr) public auth note { wards[usr] = 1; }
+    function deny(address usr) public auth note { wards[usr] = 0; }
+    modifier auth { require(wards[msg.sender] == 1); _; }
+
+    ManagerLike public manager;
+
+    constructor(address manager_)  public {
+        wards[msg.sender] = 1;
+        manager = ManagerLike(manager_);
+    }
+
+    function depend (bytes32 what, address addr) public auth {
+        if (what == "manager") { manager = ManagerLike(addr); }
+        else revert();
+    }
+
+    function balance() public {
+        (uint action, uint amount) = manager.requestAction();
+
+        if (action == manager.ActionTake()) {
+            borrowTranches(amount);
+        }
+
+        if (action == manager.ActionGive()) {
+            repayTranches(amount);
+        }
+    }
+
+    // -- Borrow Tranches ---
+    function borrowTranches(uint requestCurrency) internal  {
+        requestCurrency = borrow(manager.junior(), requestCurrency);
+
+        if (requestCurrency > 0) {
+            requestCurrency = borrow(manager.senior(), requestCurrency);
+            return;
+        }
+        revert("request amount too high");
```

Comment:
From my understanding if requestCurrency is 0 then the borrow request was entirely covered by the junior tranche, so the transaction should not revert.

And also what if the borrow request can not be entirely covered by both tranches, should the transaction then revert? If yes we need to add an additional check after borrowing from senior

Reply:
Let's discuss it with the others. I will add the revert if.